test/post_processor/test_dual_from_empirical_frequencies.py

```diff
@@ -0,0 +1,94 @@
+)
+from povm_toolbox.post_processor import (
+    POVMPostProcessor,
+    dual_from_empirical_frequencies,
+)
+from povm_toolbox.sampler import POVMSampler
+from qiskit.circuit.random import random_circuit
+from qiskit.primitives import StatevectorSampler
+from qiskit.quantum_info import (
+    DensityMatrix,
+    SparsePauliOp,
+)
+
+
+class TestDualFromEmpiricalFrequencies(TestCase):
+    """Test that we can construct optimal dual of a POVM from empirical frequencies."""
+
+    def test_not_implemented(self):
+        """Test that errors are correctly raised."""
+        _ = ClassicalShadows(n_qubit=1).definition()[(0,)]
+        _ = DensityMatrix(np.eye(2) / 2)
+        with self.assertRaises(NotImplementedError):
+            # TODO
+            raise NotImplementedError
+
+    def test_optimal_dual(self):
+        """Test that the method constructs a valid dual."""
+        qc = random_circuit(2, 1, measure=False, seed=12)
+        rng = default_rng(96568)
+        n_qubit = qc.num_qubits
+        bias = np.array([0.5, 0.25, 0.25])
+        angles = np.array(
+            [
+                [2.01757238, -1.85001671, 2.52155716, 0.45636669, 1.17175533, -0.48263278],
+                [0.0, 0.0, 1.57079633, -2.35619449, 1.57079633, -0.78539816],
+            ]
+        )
+
+        measurement = RandomizedProjectiveMeasurements(
+            n_qubit, bias=bias, angles=angles, seed_rng=rng
+        )
+        sampler = StatevectorSampler(seed=rng)
+        povm_sampler = POVMSampler(sampler=sampler)
+        job = povm_sampler.run([qc], shots=127, povm=measurement)
+        pub_result = job.result()[0]
+
+        observable = SparsePauliOp(
+            ["XI", "YI", n_qubit * "Y", n_qubit * "Z"], coeffs=[1.3, 1.2, -1, 1.4]
+        )
+
+        post_processor = POVMPostProcessor(pub_result)
+
+        post_processor.dual = dual_from_empirical_frequencies(povm_post_processor=post_processor)
+        exp_value, std = post_processor.get_expectation_value(observable)
+        self.assertAlmostEqual(exp_value, 0.9128662937761666)
+        self.assertAlmostEqual(std, 0.40472771240833644)
+
+        post_processor.dual = dual_from_empirical_frequencies(
+            povm_post_processor=post_processor,
+            loc=0,
+            bias=6,
+            ansatz=[DensityMatrix(np.eye(2) / 2), DensityMatrix(np.eye(2) / 2)],
+        )
+        exp_value, std = post_processor.get_expectation_value(observable)
+        self.assertAlmostEqual(exp_value, 0.9128662937761666)
+        self.assertAlmostEqual(std, 0.40472771240833644)
+
+        post_processor.dual = dual_from_empirical_frequencies(
+            povm_post_processor=post_processor,
+            bias=[6, 6],
+            ansatz=SparsePauliOp(["I"], coeffs=np.array([0.5])),
+        )
+        exp_value, std = post_processor.get_expectation_value(observable)
+        self.assertAlmostEqual(exp_value, 0.9128662937761666)
+        self.assertAlmostEqual(std, 0.40472771240833644)
```

Comment:
Please use sub-tests when you want to perform multiple assertion blocks: https://docs.python.org/3/library/unittest.html#distinguishing-test-iterations-using-subtests